src/TokenParser/BreakNodeParser.php

```diff
@@ -0,0 +1,78 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Twig.
+ *
+ * (c) Fabien Potencier
+ * (c) Armin Ronacher
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Twig\TokenParser;
+
+use Twig\Error\SyntaxError;
+use Twig\Node\BreakNode;
+use Twig\Token;
+
+final class BreakNodeParser extends AbstractTokenParser
+{
+    public function parse(Token $token): BreakNode
+    {
+        $stream = $this->parser->getStream();
+
+        if ($stream->test(Token::NUMBER_TYPE)) {
+            $target = $stream->getCurrent()->getValue();
+
+            if (!\is_int($target) || $target < 1) {
+                throw new SyntaxError(sprintf('Break target must be an integer > 1, got "%d".', $target), $stream->getCurrent()->getLine(), $stream->getSourceContext());
+            }
+
+            $stream->next();
+        } else {
+            $target = 1;
+        }
+
+        $stream->expect(Token::BLOCK_END_TYPE);
+
+        // make sure we are in a for-loop and the break target is valid
+        $forward = 0;
+        $loopTarget = $target;
+
+        do {
+            try {
+                $token = $stream->look($forward);
+            } catch (SyntaxError $e) { // poor mans detection of EOF stream
+                $loopTarget = -999;
+
+                break;
+            }
+
+            if ($token->test(Token::NAME_TYPE, 'for')) {
+                ++$loopTarget;
+            } else if ($token->test(Token::NAME_TYPE, 'endfor')) {
+                --$loopTarget;
+
+                if (0 === $loopTarget) {
+                    break;
+                }
+            }
+
+            ++$forward;
+        } while (true);
+
+        if (0 !== $loopTarget) {
+            throw new SyntaxError('Break tag target outside of loop.', $stream->getCurrent()->getLine(), $stream->getSourceContext());
```

Comment:
this looking forward to find how many loops get ended after the tag is very fragile to me. I think it would be better to implement such validation using a node visitor traversing the AST.

Reply:
Thanks for all the details.
If you've the time please let me know if I'm on the right track with the visitor. 
Will add more tests for sure!

Reply:
I've added the test